### PR TITLE
BUG: avoid incorrect stringdtype allocator sharing from array copies

### PR DESCRIPTION
**Pull request by ngoldbaum**
This fixes the issue noted at https://github.com/numpy/numpy/pull/25890#issuecomment-1972309561 by @mhvk.

The problem ultimately came down to passing the `allocator` to `new_stringdtype_instance` in `stringdtype_finalize_descr`.

I originally added that because I thought it had something to do with views, and I set `array_owned=2` in that case. This was wrong, you can only ever get to this code path if you're creating a new array, so it's wrong to ever share allocators if you're calling `new_stringdtype_instance`. As a result, I removed the third argument from that function and in all call sites and made it so `array_owned` can only be 0 or 1. I also had a buggy incorrect default value of `array_owned=2`, which I delete here and leave it as zero by default (i.e. not yet owned by an array).

This revealed a small bug in the stringdtype allocator implementation. Empty uninitialized strings set via `NpyString_pack_empty` would end up marked as initialized empty strings. The fix there is to delete the check for `flags==0` before setting the initialized flag in that case. As a result there are some changes to `TestImplementation`, since now empty uninitialized strings do not have the initialized and outside arena flags set.

### CODE REVIEW

File: numpy/_core/src/multiarray/item_selection.c [hunk 1]
```diff
@@ -337,14 +337,16 @@ PyArray_TakeFrom(PyArrayObject *self0, PyObject *indices0, int axis,
...
     }
 
-    Py_XDECREF(indices);
-    Py_XDECREF(self);
     if (out != NULL && out != obj) {
-        Py_INCREF(out);
-        PyArray_ResolveWritebackIfCopy(obj);
+        if (PyArray_ResolveWritebackIfCopy(obj) < 0) {
+            goto fail;
+        }
```
Comment (ngoldbaum) [inline numpy/_core/src/multiarray/item_selection.c:343]: This is an unrelated fix for a preexisting bug I ran into during debugging. The error value is discarded here and this function can error (it did in my testing) so we should have error checking.

File: numpy/_core/src/multiarray/stringdtype/static_string.c [hunk 2]
```diff
@@ -685,9 +685,6 @@ NpyString_dup(const npy_packed_static_string *in,
     _npy_static_string_u *in_u = (_npy_static_string_u *)in;
     char *in_buf = NULL;
     npy_string_arena *arena = &in_allocator->arena;
-    if (arena->buffer == NULL) {
-        return -1;
-    }
```
Comment (ngoldbaum) [inline numpy/_core/src/multiarray/stringdtype/static_string.c:690]: this check is unnecessary, all the code below it can run even if `arena->buffer` is `NULL`

**Review by charris — APPROVED** [review]

File: numpy/_core/src/multiarray/stringdtype/static_string.c [hunk 3]
```diff
@@ -530,7 +530,7 @@ NpyString_pack_empty(npy_packed_static_string *out)
 {
     _npy_static_string_u *out_u = (_npy_static_string_u *)out;
     unsigned char *flags = &out_u->direct_buffer.size_and_flags;
-    if (*flags == 0 || *flags & NPY_STRING_OUTSIDE_ARENA) {
```
Comment (mhvk) [inline numpy/_core/src/multiarray/stringdtype/static_string.c:533]: It had bothered me that this seemed necessary - great that it is not!

**Review by mhvk — APPROVED** [review]
Sorry for the delay - this looks all good, and indeed makes things more logical!
